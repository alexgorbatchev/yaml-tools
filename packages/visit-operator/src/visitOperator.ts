import YAML from 'yaml';

type NodePath = readonly (YAML.Node<unknown> | YAML.Document<YAML.Node<unknown>> | YAML.Pair<unknown, unknown>)[];

/**
 * A utility for the [yaml](https://github.com/eemeli/yaml) to process special
 * nodes, lets call them operators, with the purpose of making plain YAML a
 * little bit smarter.
 *
 * With this visitor, you can implement YAML file inclusion or node substitution
 * for example. Visitors can return a new map, sequence, scalar or undefined.
 */
export const visitOperator = (
  doc: YAML.Document,
  operator: string,
  cb: (args: readonly string[], node: YAML.Pair, path: NodePath) => YAML.Pair | YAML.Pair[] | undefined | void,
): YAML.Document => {
  const results = doc.clone();

  YAML.visit(results, {
    Pair(key, node, path) {
      if (!YAML.isScalar(node.key) || !String(node.key.value).startsWith(operator)) return;

      const [, ...args] = String(node.key.value).slice(1).split(/\s+/);
      const result = cb(args, node, path);

      if (typeof result === 'undefined') return;

      if (YAML.isPair(result)) {
        return result;
      }

      if (Array.isArray(result)) {
        const parent = path[path.length - 1];

        if (YAML.isMap(parent)) {
          const before = parent.items.slice(0, Number(key));
          const after = parent.items.slice(Number(key) + 1);
          parent.items = [...before, ...result, ...after];
        }

        return;
      }

      const parentMap = path[path.length - 1];

      if (YAML.isMap(parentMap) && YAML.isMap(result)) {
        const additions = result.items;
        const before = parentMap.items.slice(0, Number(key));
        const after = parentMap.items.slice(Number(key) + 1);
        parentMap.items = [...before, ...additions, ...after];
        return;
      }

      const parentSeq = path[path.length - 2];

      if (YAML.isSeq(parentSeq) && YAML.isSeq(result)) {
        const index = parentSeq.items.indexOf(parentMap);
        const additions = result.items;
        const before = parentSeq.items.slice(0, index);
        const after = parentSeq.items.slice(index + 1);
        parentSeq.items = [...before, ...additions, ...after];
        return;
      }
    },
  });

  return results;
};
