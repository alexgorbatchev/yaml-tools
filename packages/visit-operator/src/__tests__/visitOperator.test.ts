import YAML from 'yaml';

import { visitOperator } from '../visitOperator.js';
import { getYamlFixture } from './helpers.js';

test('visitOperator', async () => {
  let yaml: YAML.Document = await getYamlFixture('fixtures/visitOperator.yaml');

  yaml = visitOperator(yaml, '+dependencies', (args, node) => {
    if (!YAML.isMap(node.value)) return;
    const key = new YAML.Scalar(args[0]);
    const pair = yaml.createPair(key, node.value.items);
    return pair;
  });

  yaml = visitOperator(yaml, '+with', (_args, node) => {
    if (!YAML.isSeq(node.value)) return;
    const pairs = node.value.items.map((item, index) => yaml.createPair(item, index));
    return pairs;
  });

  expect(yaml.toString()).toMatchSnapshot();
});
