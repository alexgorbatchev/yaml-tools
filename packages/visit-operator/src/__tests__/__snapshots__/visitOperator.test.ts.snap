// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitOperator 1`] = `
"deps:
  linting:
    - linting-1: ^2.4.2
    - linting-2: ^5.8.0
    - linting-3: ^8.5.0
  testing:
    - testing-1: ^27.0.3
    - # comments should be ok
      testing-2: ^27.4.5
    - testing-3: ^27.1.2

here:
  too:
    nested:
      - nested-1: ^16.11.6
      - nested-2: next

# comments
results:
  set1:
    nested: 0
    testing: 1
    pkg: latest

  set2:
    nested: 0
"
`;
